StackOverflowError when method returns javax.ws.rs.core.Response

I have a simple POJO which the login method returns a Response object.  

import javax.ws.rs.Consumes;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Response;

@Produces("application/json")
@Consumes("application/json")
@Path("/auth")
public interface AuthenticationWS {

```
@GET
@Path("/login")
public Response login();

@GET
@Path("/logout")
public void logout();
```

}

and the plugin failed with a StackOverflowError

[ERROR] Exit code: 1 - javadoc: error - In doclet class com.yammer.dropwizard.apidocs.ServiceDoclet,  method start has thrown an exception java.lang.reflect.InvocationTargetException
[ERROR] java.lang.StackOverflowError
[ERROR] at com.sun.tools.javadoc.AbstractTypeImpl.<init>(AbstractTypeImpl.java:31)
[ERROR] at com.sun.tools.javadoc.WildcardTypeImpl.<init>(WildcardTypeImpl.java:32)
[ERROR] at com.sun.tools.javadoc.TypeMaker.getType(TypeMaker.java:50)
[ERROR] at com.sun.tools.javadoc.TypeMaker.getType(TypeMaker.java:29)
[ERROR] at com.sun.tools.javadoc.TypeMaker.getTypes(TypeMaker.java:85)
[ERROR] at com.sun.tools.javadoc.TypeMaker.getTypes(TypeMaker.java:75)
[ERROR] at com.sun.tools.javadoc.ParameterizedTypeImpl.typeArguments(ParameterizedTypeImpl.java:50)
[ERROR] at com.yammer.dropwizard.apidocs.ServiceDoclet.parseModels(ServiceDoclet.java:296)
[ERROR] at com.yammer.dropwizard.apidocs.ServiceDoclet.parseModels(ServiceDoclet.java:300)
[ERROR] at com.yammer.dropwizard.apidocs.ServiceDoclet.parseModels(ServiceDoclet.java:300)

If I remove the Response object, it works.  How can I get around this?  

Thanks
Arthur

--- a/README.md
+++ b/README.md
@@ -36,12 +36,12 @@ To use the Swagger Doclet in your Maven project, add the following to your POM f
                             <doclet>ServiceDoclet</doclet>
                             <docletArtifact>
                                 <groupId>com.unclehulka</groupId>
-                                <artifactId>swagger-jaxrs-doclet</artifactId>
-                                <version>0.0.1.6-SNAPSHOT</version>
+                                <artifactId>jaxrs-doclet</artifactId>
+                                <version>0.0.2-SNAPSHOT</version>
                             </docletArtifact>
                             <reportOutputDirectory>${project.build.outputDirectory}</reportOutputDirectory>
                             <useStandardDocletOptions>false</useStandardDocletOptions>
-                            <additionalparam>-apiVersion 1 -docBasePath ${server.url} -apiBasePath ${server.url}</additionalparam>
+                            <additionalparam>-apiVersion 1 -docBasePath ${server.url}/apidocs -apiBasePath ${server.url}</additionalparam>
                         </configuration>
                         <goals>
                             <goal>javadoc</goal>

--- a/jaxrs-doclet-sample-dropwizard/src/main/java/com/yammer/dropwizard/apidocs/sample/api/Recursive.java
+++ b/jaxrs-doclet-sample-dropwizard/src/main/java/com/yammer/dropwizard/apidocs/sample/api/Recursive.java
@@ -1,0 +1,15 @@
+package com.yammer.dropwizard.apidocs.sample.api;
+
+public class Recursive {
+
+    private Recursive recursive;
+
+    public Recursive getRecursive() {
+        return recursive;
+    }
+
+    public void setRecursive(Recursive recursive) {
+        this.recursive = recursive;
+    }
+
+}

--- a/jaxrs-doclet-sample-dropwizard/src/main/java/com/yammer/dropwizard/apidocs/sample/resources/RecursiveResource.java
+++ b/jaxrs-doclet-sample-dropwizard/src/main/java/com/yammer/dropwizard/apidocs/sample/resources/RecursiveResource.java
@@ -1,0 +1,14 @@
+package com.yammer.dropwizard.apidocs.sample.resources;
+
+import com.yammer.dropwizard.apidocs.sample.api.Recursive;
+
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+@Path("/Recursive")
+public class RecursiveResource {
+    @POST
+    public Recursive recurse(Recursive recursive) {
+        return recursive;
+    }
+}

--- a/jaxrs-doclet/src/main/java/com/yammer/dropwizard/apidocs/ServiceDoclet.java
+++ b/jaxrs-doclet/src/main/java/com/yammer/dropwizard/apidocs/ServiceDoclet.java
@@ -127,8 +127,7 @@ public class ServiceDoclet {
                             //build model for body parameter
                             for (Parameter parameter : method.parameters()) {
                                 if (shouldIncludeParameter(parameter)) {
-                                    if (paramTypeOf(parameter).equalsIgnoreCase("body") && !PRIMITIVES.contains(typeOf(parameter.type())))
-                                        classModelMap = parseModels(parameter.type(), classModelMap);
+                                    classModelMap = parseModels(parameter.type(), classModelMap);
                                 }
                             }
 
@@ -139,11 +138,8 @@ public class ServiceDoclet {
                                 if (me.getReturnType() == null || !me.getReturnType().equals(name)) {
                                     me.setReturnType(name);
                                 }
-                                if (!PRIMITIVES.contains(name)) {
-                                    classModelMap = parseModels(type, classModelMap);
-                                }
+                                classModelMap = parseModels(type, classModelMap);
                             }
-
                         }
                     }
                     apiMap.put(apiPath, methodMap);
@@ -261,63 +257,75 @@ public class ServiceDoclet {
      */
     private static Map<String, Model> parseModels(Type type, Map<String, Model> modelMap) {
         String typeName = typeOf(type);
-        ClassDoc cd = type.asClassDoc();
-        if (cd != null) {
-            Model model = modelMap.get(typeName);
-            if (model == null) {
-                Map<String, Type> eleMap = new HashMap<>();
+        if(!PRIMITIVES.contains(typeName) && !type.qualifiedTypeName().startsWith("javax.")) {
+            ClassDoc cd = type.asClassDoc();
+            if (cd != null) {
+                Model model = modelMap.get(typeName);
+                if (model == null) {
+                    Map<String, Type> eleMap = new HashMap<>();
 
-                //Get fields
-                FieldDoc[] fdArr = cd.fields();
-                if (fdArr != null && fdArr.length > 0) {
-                    for (FieldDoc fd : fdArr) {
-                        if (eleMap.get(fd.name()) == null) {
-                            eleMap.put(fd.name(), fd.type());
-                        }
-                    }
-                }
-
-                //Get methods
-                MethodDoc[] mdArr = cd.methods();
-                if (mdArr != null && mdArr.length > 0) {
-                    for (MethodDoc md : mdArr) {
-                        if (md.name().startsWith("get") && md.name().length() > 3) {
-                            String name = md.name().substring(3);
-                            name = name.substring(0, 1).toLowerCase() + (name.length() > 1 ? name.substring(1) : "");
-                            if (eleMap.get(name) == null) {
-                                eleMap.put(name, md.returnType());
+                    //Get fields
+                    FieldDoc[] fdArr = cd.fields();
+                    if (fdArr != null && fdArr.length > 0) {
+                        for (FieldDoc fd : fdArr) {
+                            if (eleMap.get(fd.name()) == null) {
+                                eleMap.put(fd.name(), fd.type());
                             }
                         }
                     }
-                }
 
-                //Process all fields & methods
-                if (eleMap.keySet().size() > 0) {
-                    Map<String, Property> fieldMap = new HashMap<>();
-                    for (Map.Entry<String, Type> entry : eleMap.entrySet()) {
-                        //Check if it is a collection and get collection type
-                        String containerOf = null;
-                        ParameterizedType pt = entry.getValue().asParameterizedType();
-                        if (pt != null) {
-                            Type[] typeArgs = pt.typeArguments();
-                            if (typeArgs != null && typeArgs.length > 0) {
-                                containerOf = typeOf(typeArgs[0]);
-                                if (!PRIMITIVES.contains(containerOf)) {
-                                    parseModels(typeArgs[0], modelMap);
+                    //Get methods
+                    MethodDoc[] mdArr = cd.methods();
+                    if (mdArr != null && mdArr.length > 0) {
+                        for (MethodDoc md : mdArr) {
+                            if (md.name().startsWith("get") && md.name().length() > 3) {
+                                String name = md.name().substring(3);
+                                name = name.substring(0, 1).toLowerCase() + (name.length() > 1 ? name.substring(1) : "");
+                                if (eleMap.get(name) == null) {
+                                    eleMap.put(name, md.returnType());
                                 }
                             }
                         }
+                    }
 
-                        //Add to map
-                        String eleTypeName = typeOf(entry.getValue());
-                        fieldMap.put(entry.getKey(), new Property(eleTypeName, null, containerOf));
+                    //Process all fields & methods
+                    if (eleMap.keySet().size() > 0) {
 
-                        //If not primitive, build the model for it too
-                        if (!PRIMITIVES.contains(eleTypeName)) {
+                        //Add to the model
+                        Map<String, Property> fieldMap = new HashMap<>();
+                        for (Map.Entry<String, Type> entry : eleMap.entrySet()) {
+                            //Check if it is a collection and get collection type
+                            String containerOf = null;
+                            ParameterizedType pt = entry.getValue().asParameterizedType();
+                            if (pt != null) {
+                                Type[] typeArgs = pt.typeArguments();
+                                if (typeArgs != null && typeArgs.length > 0) {
+                                    containerOf = typeOf(typeArgs[0]);
+                                }
+                            }
+
+                            //Add to map
+                            String eleTypeName = typeOf(entry.getValue());
+                            fieldMap.put(entry.getKey(), new Property(eleTypeName, null, containerOf));
+
+                        }
+
+                        modelMap.put(typeName, new Model(typeName, fieldMap));
+
+                        //Build contained models
+                        for (Map.Entry<String, Type> entry : eleMap.entrySet()) {
+                            //Check if it is a collection and get collection type
+                            ParameterizedType pt = entry.getValue().asParameterizedType();
+                            if (pt != null) {
+                                Type[] typeArgs = pt.typeArguments();
+                                if (typeArgs != null && typeArgs.length > 0) {
+                                    parseModels(typeArgs[0], modelMap);
+                                }
+                            }
+
                             parseModels(entry.getValue(), modelMap);
                         }
                     }
-                    modelMap.put(typeName, new Model(typeName, fieldMap));
                 }
             }
         }
@@ -331,8 +339,13 @@ public class ServiceDoclet {
         AnnotationDesc[] annotations = parameter.annotations();
 
         //Include if it has no annotations
-        if (annotations == null || annotations.length == 0)
-            return true;
+        if (annotations == null || annotations.length == 0){
+            if (paramTypeOf(parameter).equalsIgnoreCase("body")){
+                return true;	
+            } else {
+                return false;
+            }
+        }
 
         //Include if it has a jax-rs annotation
         for (AnnotationDesc annotation : annotations) {

